### PR TITLE
[camera] Add implementations for `camera_platform_interface` package.

### PR DESCRIPTION
**Pull request by BeMacized**
## Description

Updates the dart plugin and its native iOS and Android implementations to make use of the new platform interface that has been submitted in #3253

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [ ] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by mvanbeusekom** [comment]
@googlebot I consent.

**Comment by danielroek** [comment]
@googlebot I consent.

**Comment by ditman** [comment]
(I've published camera_platform_interface, sorry for the delay!)

**Comment by BeMacized** [comment]
> (I've published camera_platform_interface, sorry for the delay!)

Nice! The dependency should now be updated as well.

**Comment by ditman** [comment]
Hey @BeMacized @mvanbeusekom, can you edit this PR so I'm allowed to push changes to it (and its branch?) I have a couple of fixes for the issues outlined by the `publishable` check that I'd like to push! I'll also add some extra comments to the code :)

**Comment by mvanbeusekom** [comment]
> Hey @BeMacized @mvanbeusekom, can you edit this PR so I'm allowed to push changes to it (and its branch?) I have a couple of fixes for the issues outlined by the `publishable` check that I'd like to push! I'll also add some extra comments to the code :)

Hi @ditman, I gave you write access to our fork of the plugins repo (https://github.com/baseflow/flutter-plugins) which should give you the required access to push changes to the branch (which will show up in the PR). You should have received an invite by email. Thanks for helping out!

File: packages/camera/camera/lib/src/camera_controller.dart
Comment (ditman) [inline packages/camera/camera/lib/src/camera_controller.dart:6]: If possible, please don't import `dart:io`, otherwise this won't be marked as a "web-compatible" plugin.

File: packages/camera/camera/lib/src/camera_controller.dart
Comment (ditman) [inline packages/camera/camera/lib/src/camera_controller.dart:282]: Please, use `defaultTargetPlatform` from Flutter instead of `Platform` from `dart:io` to do these checks (there's one here, and another one in the `stopImageStream` method)

File: packages/camera/camera/lib/src/camera_controller.dart
Comment (ditman) [inline packages/camera/camera/lib/src/camera_controller.dart:487]: This is the only usage of the _creatingCompleter, which seems to be only checking if the initalize method has been called (but not finished running). I've modified this by a `FutureOr`, it's probable this can be simplified even further...

File: packages/camera/camera/test/camera_test.dart
Comment (ditman) [inline packages/camera/camera/test/camera_test.dart:308]: Would it make sense to split the start/stop ImageStream tests to their own files, so we don't mix the PlatformInterface-based initialization and tests, with the MethodChannel ones?

Comment (ditman) [reply packages/camera/camera/test/camera_test.dart:308]: (Also for later cleanup, it's easier, just deleting a file)

**Review by ditman — COMMENTED** [review]
Hey, I've checked this (except for the Java bits, maybe @bparrishMines can take a look at those), and I have a few comments. I don't think there's anything blocking. Some things are not very idiomatic. My only major concerns:

* Reexporting the whole camera_platform_interface from camera. Please reexport only the things that are being removed from Camera. That way we can expose things on platform_interface that are useful to other platform authors, but that shouldn't be directly seen by the users of the plugin.
* Do not import `dart:io`, that way the plugin will be marked as "WEB" compatible in pub.dev later!

**Comment by ditman** [comment]
@googlebot I consent.

**Comment by mvanbeusekom** [comment]
@googlebot I consent.

**Comment by ditman** [comment]
> Hi @ditman, I gave you write access to our fork of the plugins repo

I have added a couple of commits above, I don't think I've broken anything (tests seemed to continue passing), but please verify those! They have to do with the initialization sequence and how it used completers a little bit weirdly.

(Also, when creating PRs, I think it's enough to mark the "allow maintainers to do changes to this PR" checkbox to enable me to push stuff to the branch, without having to give me access to the whole repo (but thanks anyway, Maurits!))

File: packages/camera/camera/lib/src/camera_controller.dart
Comment (mvanbeusekom) [reply packages/camera/camera/lib/src/camera_controller.dart]: Done!

Comment (ditman) [reply packages/camera/camera/lib/src/camera_controller.dart]: Thanks!!

File: packages/camera/camera/lib/src/camera_controller.dart
Comment (mvanbeusekom) [reply packages/camera/camera/lib/src/camera_controller.dart]: Yes that looks much better. Thanks

File: packages/camera/camera/lib/src/camera_controller.dart
Comment (mvanbeusekom) [reply packages/camera/camera/lib/src/camera_controller.dart]: Perfect. I agree this is the better solution

File: packages/camera/camera/lib/camera.dart
Comment (mvanbeusekom) [reply packages/camera/camera/lib/camera.dart:9]: I seem to have some problems with this one. Locally I have updated this statement to:

```dart
export 'package:camera_platform_interface/camera_platform_interface.dart'
    show
        CameraDescription,
        CameraException,
        CameraLensDirection,
        ResolutionPreset,
        XFile;
```

But now the code in the example/lib/main.dart won't compile anymore because it can't locate any of the types mentioned by the `show` statement. I made sure I imported the `import package:camera/camera.dart` Dart file but still not working.

I can't figure out what I am doing wrong here, probably overlooking something (it is getting late here). Maybe you have some pointers? Will give it a fresh look tomorrow morning.

Comment (ditman) [reply packages/camera/camera/lib/camera.dart:9]: It should work, maybe you need a flutter clean in the example app, or to "restart" the analyzer (or both?).

If you want, push whatever you have to the branch and I can take a look on my end (it's going to break some tests, but oh well!)

File: packages/camera/camera/test/camera_test.dart
Comment (mvanbeusekom) [reply packages/camera/camera/test/camera_test.dart]: I agree, thanks for pointing it out. I moved the tests into a new file called `camera_image_stream_test.dart`.

File: packages/camera/camera/lib/camera.dart
Comment (mvanbeusekom) [reply packages/camera/camera/lib/camera.dart]: it was indeed my environment that got a bit confused ;) I corrected everything and it works without issues now.

**Comment by mvanbeusekom** [comment]
@ditman I have processed your feedback, would really appreciate it if you can have another look. 

The changes you made look good, I tested the app on several Android and iOS devices and still works nicely.

File: packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/DartMessengerTest.java
Comment (bparrishMines) [inline packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/DartMessengerTest.java:56]: nit: We typically prefer using literals in tests over constants. It makes it easier to read and self contained. (e.g. "error" instead of `DartMessenger.EventType.ERROR.toString().toLowerCase()`). Up to you though.

File: packages/camera/camera/ios/Classes/CameraPlugin.m
Comment (bparrishMines) [inline packages/camera/camera/ios/Classes/CameraPlugin.m:385]: nit: I think `@"error"` would be better as a constant like it is in Java so it doesn't need to be copy and pasted everywhere.

**Review by bparrishMines — COMMENTED** [review]
Platform code look good to me! I left a couple of small nits. I'll also let @ditman give the approval once he believes it's ready.

File: packages/camera/camera/ios/Classes/CameraPlugin.m
Comment (mvanbeusekom) [reply packages/camera/camera/ios/Classes/CameraPlugin.m]: Agreed, updated the code to use a constant called `errorMethod` and use it where we do a `invokeMethod:@"error"`.

**Comment by mvanbeusekom** [comment]
@ditman, @bparrishMines, I have updated the code to include the feedback from Maurice. @ditman would be great if you can give it another look.

File: packages/camera/camera/lib/camera.dart
Comment (ditman) [inline packages/camera/camera/lib/camera.dart:15]: Compare how much cleaner this is now with everything that is exported by the platform_interface:
https://pub.dev/documentation/camera_platform_interface/latest/camera_platform_interface/camera_platform_interface-library.html! Thanks!!

**Review by ditman — APPROVED** [review]
I think this is good to go, however the analysis is reporting some formatting errors. Let's merge this after we get a full green CI run!

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java
Comment (hamdikahloun) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:226]: Can we handle  [SecurityException](https://developer.android.com/reference/java/io/File#createTempFile(java.lang.String,%2520java.lang.String)) too ?
I see we have static file name for pictures. We can save one picture?

Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:226]: This is not the case, the `createTempFile` method will create a unique file name. The first parameter ("CAP" in this case) is a prefix that will be used to generate the name (see [File.createTempFile](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#createTempFile(java.lang.String,%20java.lang.String,%20java.io.File) for details).

The actual file will look something like "CAP3602253894598046604.jpg"

Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:226]: Good catch regarding the `SecurityException`, I will add support to handle this as well.

Comment (mvanbeusekom) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:226]: done

Comment (hamdikahloun) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/Camera.java:226]: Great

**Review by hamdikahloun — APPROVED** [review]
LGTM

**Comment by AlexV525** [comment]
> * [x]  Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).

I just wondering why only the **MINOR** version has been increased?

**Comment by mvanbeusekom** [comment]
@AlexV525  it actually is a major version number. When a package is not considered stable (version 1.0.0) Dart considers the second digit the major number, the third digit as minor and a fourth digit (added as +n) as the patch number: 

> Although semantic versioning doesn’t promise any compatibility between versions prior to 1.0.0, the Dart community convention is to treat those versions semantically as well. The interpretation of each number is just shifted down one slot: going from 0.1.2 to 0.2.0 indicates a breaking change, going to 0.1.3 indicates a new feature, and going to 0.1.2+1 indicates a change that doesn’t affect the public API. For simplicity’s sake, avoid using + after the version reaches 1.0.0.

The quote above is copied from the article [Package versioning | Dart](https://dart.dev/tools/pub/versioning), in particular the [Semantic versions](https://dart.dev/tools/pub/versioning#semantic-versions) chapter.

**Comment by AlexV525** [comment]
https://github.com/flutter/flutter/wiki/Package-migration-to-1.0.0 This might be against to the migration proposal. The package using `>=0.x.y+z <2.0.0` directly suffered from breaking changes.

**Comment by mvanbeusekom** [comment]
Not really sure how to answer this in a satisfying manor ;). The Flutter plugin teams follows the versioning semantics as discussed in the [Package versioning | Dart](https://dart.dev/tools/pub/versioning) article. As such we bumped the second digit to indicate a major version change.

**Comment by AlexV525** [comment]
Thanks for the response. I think I'll consult others for the conflict later, since they're obviously against each other at present.

**Pull request merged**
